Gossiper: add endpoint_id to interface

bhalevy opened the pull request:
This series adds an `endpoint_id` to the gossiper interface as the first step to basing its interface solely on `host_id`.

struct `endpoint_id` describes the endpoint as a pair of `{ host_id, inet_address }`.
It allows the notification listeners to use the `host_id` when available or to to use the `inet_address` when required (until other layers in the stack, like the `token_metadata` are converted), or as a fallback.

The RPC `client_info` is used to determine the peer's host_id.
Plus, the gossiper was changed to always send the HOST_ID application state along in the endpoint_state in the ack and ack2 responses, but when talking to nodes running an older version this is not guranteed, hence the host_id cannot be mandated yet.

An address map, mutally mapping host_id to and from endpoints' inet_address on all shards, was added.
It is needed to keep the mappings learnt in the shadow round.
The raft address map was considered too, but it is insufficient for the gossiper at this point since:
a. it's available only on shard 0, and
b. the raft address map is updated by callbacks called from gossiper, while the latter needs the mapping earlier than that.

The `_endpoint_state_map` was extended by `_endpoint_state_map_by_host_id` providing fast lookup of the endpoint_state wither by the endpoint address or by its host_id.
This serves a nasty use case of replacing a node with the same ip address.
In the future it may be possible to get rid of getting the endpoint state by inet_address and
a clean api based solely on host_id could be established, once we no longer have to search for the endpoint state by its ip address.

Refs https://github.com/scylladb/scylladb/issues/12282

bhalevy commented:
cc @kostja @kbr-scylla @gleb-cloudius

bhalevy commented:
Cc @avikivity

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3721/

bhalevy commented:
Depends also on https://github.com/scylladb/scylladb/pull/15459

gleb-cloudius commented on main.cc:
```diff
@@ -1917,6 +1918,12 @@ To start the scylla server proper, simply invoke as: scylla server (or just scyl
             // just ignore it and exit normally
             _exit(0);
             return 0;
+          } catch (const abort_requested_exception&) {
```
It looks like indentation is broken throughout this code.

bhalevy replied:
unless you accept 2-characters wide indentation.
in any case, this is orthogonal, and sent separately as https://github.com/scylladb/scylladb/pull/15443

gleb-cloudius replied:
Yes, it is pre-existing issue.

gleb-cloudius commented on gms/gossiper.cc:
```diff
@@ -518,7 +518,8 @@ void gossiper::init_messaging_service_handler() {
         return handle_echo_msg(from, generation_number_opt);
     });
-    _messaging.register_gossip_shutdown([this] (inet_address from, rpc::optional<int64_t> generation_number_opt) {
+    _messaging.register_gossip_shutdown([this] (const rpc::client_info& cinfo, gms::inet_address, rpc::optional<int64_t> generation_number_opt) {
```
> messaging, gossiper: pass client_info to shutdown message handler

Just to log it? Can you add a motivation to the commit message?

bhalevy replied:
Sure.  The motivation is to convey the sender's host_id in it.

gleb-cloudius replied:
Current verb does not rely on `rpc::client_info` to get a source. The id of the node that is been shutting down is sent as one of the parameters. So you want to change that and take the info from `rpc::client_info` instead?

bhalevy replied:
But we send the endpoint `inet_address`.
I'd like to take the info from `client_info` to get the host_id from there, assuming the shutdown message is always sent from the shutting down node, not on behalf of another node.

gleb-cloudius commented on gms/gossiper.hh:
```diff
@@ -104,11 +104,11 @@ private:
+    future<> handle_ack_msg(const rpc::client_info& cinfo, gossip_digest_ack ack_msg);
+    future<> handle_ack2_msg(const rpc::client_info& cinfo, gossip_digest_ack2 msg);
+    future<> handle_echo_msg(const rpc::client_info& cinfo, std::optional<int64_t> generation_number_opt);
+    future<> handle_shutdown_msg(const rpc::client_info& cinfo, std::optional<int64_t> generation_number_opt);
```
Isn't it time to move gossiper verbs to the IDL?

bhalevy replied:
Yeah, it could be done as a preliminary cleanup PR

tgrabiec commented:
> The raft address map was considered too, but it is insufficient for the gossiper at this point since:
> a. it's available only on shard 0, and

raft_address_map is replicated to all shards, see raft_address_map_t::replicate().

> b. the raft address map is updated by callbacks called from gossiper, while the latter needs the mapping earlier than that.

We can initialize raft_address_map earlier.

bhalevy commented:
> > The raft address map was considered too, but it is insufficient for the gossiper at this point since:
> > a. it's available only on shard 0, and
> 
> raft_address_map is replicated to all shards, see raft_address_map_t::replicate().

Why is it done in the background by the way?

> > b. the raft address map is updated by callbacks called from gossiper, while the latter needs the mapping earlier than that.
> 
> We can initialize raft_address_map earlier.

I'll take another stab at it

gleb-cloudius commented on gms/gossiper.cc:
```diff
@@ -441,12 +448,12 @@ future<> gossiper::handle_echo_msg(const rpc::client_info& cinfo, std::optional<
 }
 
 future<> gossiper::handle_shutdown_msg(const rpc::client_info& cinfo, std::optional<int64_t> generation_number_opt) {
+    co_await update_address_map(cinfo);
```
Do we really need to update the map on each and every gossiper messages? Here for instance a node reports that it is going to shutdown. It may boot with another IP. Why is it important to update the map here? Shouldn't we update it only on a first message the gossiper sends (echo?)?

bhalevy replied:
If the peer behaves itself, it should suffice.
Updating the address_map is supposed to be lightweight and apply only when there is no such mapping.
But with using the raft_address_map we can reconsider that and update only on echo or syn (which is sent also to unreachable endpoints).

bhalevy commented:
In [ab8f970](https://github.com/scylladb/scylladb/commit/ab8f970d01f07ad2fb6fb170750831e5c0bfa793):
- gossiper: get_or_create_endpoint_state: create empty endpoint_state
- use raft address map
  - return future from add_or_update_entry at el.
- update_address_map only from echo and syn handlers

tgrabiec commented:
> > > The raft address map was considered too, but it is insufficient for the gossiper at this point since:
> > > a. it's available only on shard 0, and
> > 
> > 
> > raft_address_map is replicated to all shards, see raft_address_map_t::replicate().
> 
> Why is it done in the background by the way?

Not sure. @gleb-cloudius @kostja ?

kbr-scylla commented:
> Why is it done in the background by the way?

Because of the interface. Most of `raft_address_map`'s API don't return a `future`. And replication in background was enough to satisfy the requirements we had for original use cases of `raft_address_map`. See https://github.com/scylladb/scylladb/pull/11791#discussion_r999241018

bhalevy commented on gms/gossiper.cc:
```diff
@@ -166,6 +169,7 @@ void gossiper::do_sort(utils::chunked_vector<gossip_digest>& g_digest_list) cons
 // Depends on
 // - no external dependency
 future<> gossiper::handle_syn_msg(const rpc::client_info& cinfo, gossip_digest_syn syn_msg) {
+    co_await update_address_map(cinfo);
```
@kbr-scylla but the generation comes from gossiper endpoint_state, and here we don't have that in the client_info.
I like @tgrabiec's idea of fencing off messages from stale peer.
The quesation is would it need adding connection establishment verbs or not.

tgrabiec replied:
> The quesation is would it need adding connection establishment verbs or not.

It would fit into messaging_verb::CLIENT_ID, but it is no_wait, so we can't reuse it. We need to add SERVER_ID.

tgrabiec replied:
SERVER_ID could also return the generation which we can use when populating the address mapping.

tgrabiec replied:
> If there's an existing connection for that host_id, wouldn't it be better to disconnect it first before allowing a new connection?

There may be multiple connections (from different shards) coming from a single peer.

bhalevy replied:
> > If there's an existing connection for that host_id, wouldn't it be better to disconnect it first before allowing a new connection?
> 
> There may be multiple connections (from different shards) coming from a single peer.

OK, makes sense.
How about a stale connection with same host_id and different inet_address?
Is it even possible, or since it requires restart the previous connections are bound to be closed already?

tgrabiec replied:
> How about a stale connection with same host_id and different inet_address?
> Is it even possible, or since it requires restart the previous connections are bound to be closed already?

I think it's possible if the old node did not close the connection gracefully. 

We could close these, but what guarantee would you like to achieve by that?

bhalevy replied:
I'd like to fence possibly in-flight updates from the previous incarnation.
That said, ordering by version/generation should make sure that older updates to the endpoint state are ignored.

tgrabiec replied:
> > It would fit into messaging_verb::CLIENT_ID, but it is no_wait, so we can't reuse it.
> 
> Do we need the establisher of connection (sender) to receive a response?

For host_id-based routing yes, because we want to know which host we connected to.

bhalevy commented on gms/gossiper.cc:
```diff
@@ -1416,6 +1425,67 @@ endpoint_state_ptr gossiper::get_endpoint_state_ptr(inet_address ep) const noexc
+                } catch (...) {
+                    if (prev) {
+                        it->second = *prev;
+                    }
```
this code is gone in the latest version that reuses the raft_address_map.

bhalevy commented on gms/gossiper.cc:
```diff
@@ -1416,6 +1425,67 @@ endpoint_state_ptr gossiper::get_endpoint_state_ptr(inet_address ep) const noexc
+                    throw;
+                }
+                if (prev) {
+                    g._address_to_host_id_map.erase(*prev);
```
core is gone in latest version

bhalevy commented on gms/gossiper.cc:
```diff
@@ -571,12 +571,32 @@ future<> gossiper::send_gossip(gossip_digest_syn message, std::set<inet_address>
+            if (remote_host_id) {
+                if (local_host_id != remote_host_id) {
+                    logger.warn("Node {} changed host_id from {} to {}", addr, local_host_id, remote_host_id);
+                    co_return co_await this->handle_major_state_change(node, std::move(remote_state), permit.id());
```
When searching using the endpoint this would be the "replace with same ip address" case (and the log message sucks, need to improve it.). Once we search by host_id first we should find no endpoint_state for the new node, so `local_host_id != remote_host_id` should never happen and would indicate a bug.

tgrabiec replied:
Yes, so by the end of the series, this should be impossible. If we don't allow null host ids, this will be explicit in the code and there will be no need for code like this.

bhalevy commented on gms/gossiper.cc:
```diff
@@ -1425,14 +1420,16 @@ endpoint_state_ptr gossiper::get_endpoint_state_ptr(inet_address ep) const noexc
-        return update_address_map(host_id, addr);
+        co_await update_address_map(host_id, addr);
+    } else {
+        host_id = get_host_id(addr, throw_on_error::no);
```
I'm not sure we can trust nodes running old versions to always have the HOST_ID in their application state.

tgrabiec replied:
It's set together with everything else in prepare_to_join(), and the line dates back to 2015:
```
    app_states.emplace(gms::application_state::HOST_ID, value_factory.host_id(local_host_id));
```
So it's pretty much there since the beginning.

bhalevy replied:
Maybe I needed it due to a bug I had in the series with updating the endpoint_state.
I'll check if we can mandate a valid host_id here.

tgrabiec commented on gms/gossiper.hh:
```diff
@@ -215,6 +215,7 @@ private:
 
     /* map where key is the endpoint and value is the state associated with the endpoint */
     std::unordered_map<inet_address, endpoint_state_ptr> _endpoint_state_map;
+    std::unordered_map<locator::host_id, endpoint_state_ptr> _endpoint_state_map_by_host_id;
```
We should be able to have a single map indexed by host_id. If you don't have a host id, you're not allowed to insert states. That will simplify things.

tgrabiec commented on locator/topology.cc:
```diff
@@ -438,6 +438,29 @@ const node* topology::find_node(node::idx_type idx) const noexcept {
+    } else {
+        if (!opt_ep) {
+            on_internal_error(tlogger, format("add_or_update_endpoint: host_id={}: opt_ep must be engaged", id));
+        }
```
There is a bug in raft topology, where after decommissioning a node, you're not able to restart nodes because they fail to resolve IPs of left nodes.

tgrabiec replied:
We need to identify purely by UUID to fix this.

bhalevy commented on gms/gossiper.cc:
```diff
@@ -569,12 +569,12 @@ future<> gossiper::send_gossip(gossip_digest_syn message, std::set<inet_address>
     locator::host_id remote_host_id = remote_state.get_host_id();
     locator::host_id local_host_id;
-    const auto& addr = node;
+    auto node = endpoint_id(remote_host_id ? remote_host_id : get_host_id(addr), addr);
```
By the end of the series, with "gms: endpoint_state_map: maintain state_by_host_id map" we lookup the endpoint_state internally by the host_id, when available, so we do update it incrementally after ip address change.
I'll see if we we can index the endpoint_state only by host_id and rely on the address map to translate ip -> host_id.
Last time I tried that all hell broke loose...
To have a pure hsot_id interface I'm afraid we'll need to change the gossip protocol to use the host_id to identify the endpoints and maybe add the broadcast address as an application state (I was mistaken to think it's the `RPC_ADDRESS` but later realized it's a different RPC).

kbr-scylla commented on gms/gossiper.cc:
```diff
@@ -1500,12 +1526,35 @@ const endpoint_state& gossiper::get_endpoint_state(inet_address ep) const {
+    } catch (...) {
+        if (prev) {
+            it->second = std::move(prev);
+        }
```
Should we remove from `_endpoint_state_map_by_host_id` if `inserted`?

bhalevy replied:
yes, thanks.

bhalevy commented on gms/gossiper.cc:
```diff
@@ -387,8 +389,9 @@ future<> gossiper::do_send_ack2_msg(msg_addr from, utils::chunked_vector<gossip_
 // - on_alive callbacks
-future<> gossiper::handle_ack2_msg(msg_addr from, gossip_digest_ack2 msg) {
-    logger.trace("handle_ack2_msg():msg={}", msg);
+future<> gossiper::handle_ack2_msg(const rpc::client_info& cinfo, gossip_digest_ack2 msg) {
```
> In later patch you call `get_endpoint_id(cinfo)`. Maybe we could do that before `background_msg`, and pass `endpoint_id` to `background_msg` call instead of `client_info`?

Sounds like a good idea.  Will look into it.

bhalevy commented on gms/gossiper.cc:
```diff
@@ -422,13 +422,19 @@ future<> gossiper::handle_echo_msg(msg_addr from, std::optional<int64_t> generat
+                    logger.log(respond ? log_level::debug : log_level::warn, "handle_echo_msg: from={}, saved_generation_number={}, current_generation_number={}",
                             from, saved_generation_number, current_generation_number);
+                } else {
+                    logger.warn("handle_echo_msg: endpoint_state not found for node {}", from);
```
I am not sure. I don't remember ever seeing them in practice, but maybe I just haven't noticed.
I'm pretty sure they never come in bulk.

bhalevy commented on gms/gossiper.cc:
```diff
@@ -422,13 +422,19 @@ future<> gossiper::handle_echo_msg(msg_addr from, std::optional<int64_t> generat
         } else {
             respond = true;
         }
+    } else {
```
These are warnings since the node got an echo messages and it decides not respond, for whatever reason.
Maybe if the other node is quarantined or left we can even log the message in debug mode.
At any rate, this patch is immaterial to the PR so I'm thinking about taking it out and submitting in a followup series.

bhalevy commented on gms/gossiper.cc:
```diff
@@ -1466,6 +1466,15 @@ stop_iteration gossiper::for_each_endpoint_state_until(std::function<stop_iterat
 }
 
+future<stop_iteration> gossiper::for_each_endpoint_state_gently_until(std::function<future<stop_iteration>(const inet_address&, const endpoint_state&)> func) const {
+    for (const auto& [node, eps] : _endpoint_state_map) {
```
> But... can't _endpoint_state_map change while we're yielding?

Right. Will fix.

bhalevy commented on gms/gossiper.hh:
```diff
@@ -200,6 +200,26 @@ private:
+
+    // Update the host_id <-> address mapping on all shards
+    future<> update_address_map(const locator::host_id& host_id, inet_address addr, gms::generation_type generation_number = {}) noexcept;
+    future<> update_address_map(const rpc::client_info& cinfo) noexcept;
```
I'll just remove this overload and check the host_id validity at the caller (and add the comment there)

bhalevy commented on service/raft/raft_address_map.hh:
```diff
@@ -314,8 +314,10 @@ public:
-        rslog.warn("Finding Raft nodes by IP addresses is deprecated. Please use Host IDs instead.");
+    std::optional<raft::server_id> find_by_addr(gms::inet_address addr, bool warn = true) const {
+        if (warn) {
+            rslog.warn("Finding Raft nodes by IP addresses is deprecated. Please use Host IDs instead.");
```
It's a pre-existing warning BTW.
The call site from gossiper is new and so I added the `warn` parameter to squelch the warning when the reverse mapping may be missing in normal conditions.

bhalevy commented on gms/gossiper.cc:
```diff
@@ -2131,8 +2131,12 @@ void gossiper::build_seeds_list() {
-future<> gossiper::add_saved_endpoint(inet_address ep) {
-    if (is_me(ep)) {
+future<> gossiper::add_saved_endpoint(inet_address ep, locator::host_id host_id) {
+    if (ep == inet_address() || !host_id) {
```
We should treat a missing host_id as an error at and after this series.
You're right that the storage_service shouldn't ignore that.
`add_saved_endpoint` must not allow null host_id and its callers should not call it when they don't have the host_id in hand.

I'll make sure it's documented.

bhalevy commented:
In [e361ab5](https://github.com/scylladb/scylladb/commit/e361ab56a18ab2322ae37a55a29a4c4420b1d805):
- addressed latest @kbr-scylla review comments
- cleaned up patch series
- gms: inet_address: add bool operator
- gossiper: get_host_id: add throw_on_error param
- gossiper: pass endpoint_id to message handlers from rpc::client_info
  - rather than passing the client_info
- raft_address_map: return future from replicate and add_or_update_entry
  - copy endpoint_state `for_each_endpoint_state_gently_until` so it can yield safely
- gossiper: update address_map with client_info
  - moved call sites to callers of the respective handlers
  - open coded the logic so the address map is updated only if `host_id` is available in the `client_info`
- storage_service: keep host_id in loaded_endpoints map
  - mandate non-null host_id (previously it was optional)
- gossiper: do_apply_state_locally: call handle_major_state_change when replacing node with same ip
  - refactored out of previous patch (and the logic actually goes away in a later patch when we look up the ep_state by host_id)
- token_metadata, topology: add update_endpoint overloads using host_id
  - added doc comments in `topology::index_node`
- gossiper: refactor get_application_state helpers
- gossiper: get_endpoint_id: mandate valid host_id

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/4184/

bhalevy commented:
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/4184/

Requires https://github.com/scylladb/scylla-dtest/pull/3588
Now that this PR came out of draft mode, it would be a good time to merge the dtest changes @fruch. They should be fully backward compatible.

bhalevy commented:
https://github.com/scylladb/scylladb/commit/b33883af93f0e875c3e9d2eca7ee6c2fe6d402b8: rebased

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - Build
:x: - Sanity Tests
:x: - Unit Tests
#### Failed Tests (4/21000):
- [test_queries_during_upgrade](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4300/testReport/junit/reversed_queries_test/TestReversedQueriesSelectorsDuringUpgrade/Tests___Sanity_Tests___test_queries_during_upgrade) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_queries_during_upgrade)
- [test_increment_decrement_counters_in_threads_nodes_restarted](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4300/testReport/junit/update_cluster_layout_tests/TestUpdateClusterLayout/Tests___Sanity_Tests___test_increment_decrement_counters_in_threads_nodes_restarted) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_increment_decrement_counters_in_threads_nodes_restarted)
- [test_boot_after_ip_change](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4300/testReport/junit/%28root%29/test_boot_after_ip_change/Tests___Unit_Tests___test_boot_after_ip_change) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_boot_after_ip_change)
- [topology_custom.test_boot_after_ip_change.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4300/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_custom_test_boot_after_ip_change_debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_custom.test_boot_after_ip_change.debug.3)
#### Build Details:
- Build URL: https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/4300/
- Duration: 2 hr 40 min
- Builder: spider4.cloudius-systems.com

tgrabiec commented on gms/gossiper.cc:
```diff
@@ -1282,6 +1282,18 @@ void gossiper::make_random_gossip_digest(utils::chunked_vector<gossip_digest>& g
+    }
+    auto* host_id_state = es.get_application_state_ptr(application_state::HOST_ID);
+    if (!host_id_state) {
+        es.add_application_state(application_state::HOST_ID, versioned_value::host_id(node.host_id));
```
This comment is unresolved.

tgrabiec commented on gms/gossiper.cc:
```diff
@@ -1282,6 +1282,18 @@ void gossiper::make_random_gossip_digest(utils::chunked_vector<gossip_digest>& g
+    }
+    if (node.addr == inet_address()) {
+        on_internal_error(logger, format("replicate: node {} has null address", node));
+    }
```
It seems to me these should be validated earlier, when endpoint_state is created. replicate() is too late.

tgrabiec commented on gms/gossiper.hh:
```diff
@@ -537,10 +537,9 @@ public:
-        return endpoint_id(get_host_id(addr, throw_on_error::no), addr);
+    // Throws a runtime_error if the host is not found,
+    endpoint_id get_endpoint_id(inet_address addr) const {
+        return endpoint_id(get_host_id(addr), addr);
```
Why not squash this patch when this was introduced? We change our mind in the middle of the series, without giving a reason.

tgrabiec commented on gms/gossiper.cc:
```diff
@@ -1220,8 +1215,16 @@ int64_t gossiper::get_endpoint_downtime(inet_address ep) const noexcept {
-    if (!state || !is_alive(node)) {
+    auto state = get_endpoint_state_ptr(node.host_id);
+    if (!state) {
+        state = get_endpoint_state_ptr(node.addr);
```
Why not keep this logic inside `get_endpoint_state_ptr(node)`?

tgrabiec replied:
hmm, it's probably a bad idea to always fall back to IP. Why do we fall back to IP here?

tgrabiec commented on gms/gossiper.cc:
```diff
@@ -1804,7 +1862,7 @@ future<> gossiper::handle_major_state_change(endpoint_id node, endpoint_state ep
         }
     }
-    logger.trace("Adding endpoint state for {}, status = {}", node, get_gossip_status(eps));
+    logger.info("Adding endpoint state for {}, status = {}, at {}", node, get_gossip_status(eps), current_backtrace());
```
That will spam the logs with backtraces, shouldn't it be debug level?

tgrabiec commented on gms/gossiper.cc:
```diff
@@ -2327,6 +2388,9 @@ future<> gossiper::add_local_application_state(std::list<std::pair<application_s
             }
+            local_state.add_application_state(application_state::HOST_ID, versioned_value::host_id(node.host_id));
+            auto rpc_addr = utils::fb_utilities::get_broadcast_rpc_address();
+            local_state.add_application_state(application_state::RPC_ADDRESS, versioned_value::rpcaddress(rpc_addr));
```
Why do we need this?

This will change the two states any time a local state is changed (e.g. CACHE_HITRATES).

It's enough to add these when we add local state for the first time, which I thought we already do.

gusev-p commented:
I just wanted to clarify yet again, what are the reasons why we need this?

Now I'm moving `token_metadata` to host_ids, and it seems I can make it work. This means raft topology management will be based entirely on host_ids, and gossiper/legacy mode will be no-worse than before. We can tell users to migrate to the raft topology management to solve the problems with IP based gossiper.

@bhalevy @tgrabiec

bhalevy commented:
> I just wanted to clarify yet again, what are the reasons why we need this?
> 
> Now I'm moving `token_metadata` to host_ids, and it seems I can make it work. This means raft topology management will be based entirely on host_ids, and gossiper/legacy mode will be no-worse than before. We can tell users to migrate to the raft topology management to solve the problems with IP based gossiper.
> 
> @bhalevy @tgrabiec

Currently raft is fed with event notifications from gossip.
I'd be happy in the long term if raft will take care of the topology-related states (e.g. node STATUS, HOST_ID, DC, RACK, TOKENS/TABLETS), while we let gossip deal only with the less critical information that could propagate non-deterministically.

But even then, gossip should use host_id to identify nodes, both externally and internally because we want to use the node ip address solely for networking and not for keeping any state about the node.

bhalevy commented:
Closing this PR as it's not needed after 882f405eed8dacd433824e5d0a05221c88e9bf4b et al
